pyro/distributions/ordered_logistic.py

```diff
@@ -0,0 +1,56 @@
+import torch
+from pyro.distributions import constraints
+from pyro.distributions.torch import Categorical
+
+
+class OrderedLogistic(Categorical):
+    """
+    Alternative parametrization of the distribution over a categorical variable.
+
+    Instead of the typical parametrization of a categorical variable in terms
+    of the probability mass of the individual categories ``p``, this provides an
+    alternative that is useful in specifying ordered categorical models. This
+    accepts a vector of ``cutpoints`` which are an ordered vector of real numbers
+    denoting baseline cumulative log-odds of the individual categories, and a
+    model vector ``predictor`` which modifies the baselines for each sample
+    individually.
+
+    These cumulative log-odds are then transformed into a discrete cumulative
+    probability distribution, that is finally differenced to return the probability
+    mass matrix ``p`` that specifies the categorical distribution.
+
+    :param Tensor predictor: A tensor of predictor variables of arbitrary
+        shape. The output shape of non-batched samples from this distribution will
+        be the same shape as ``predictor``.
+    :param Tensor cutpoints: A tensor of cutpoints that are used to determine the
+        cumulative probability of each entry in ``predictor`` belonging to a
+        given category. The first `cutpoints.ndim-1` dimensions must be
+        broadcastable to ``predictor``, and the -1 dimension is monotonically
+        increasing.
+    """
+
+    arg_constraints = {
+        "predictor": constraints.real,
+        "cutpoints": constraints.ordered_vector,
+    }
+
+    def __init__(self, predictor, cutpoints, validate_args=None):
+        self.predictor = predictor
+        self.cutpoints = cutpoints
+        # calculate cumulative probability for each predictor
+        q = torch.sigmoid(cutpoints - predictor.view(predictor.shape + (1,)))
+        # turn cumulative probabilities into probability mass of categories
+        p_shape = q.shape[:-1] + (q.shape[-1] + 1,)
+        p = torch.zeros(p_shape, dtype=q.dtype, device=q.device)
+        p[..., 0] = q[..., 0]
+        p[..., 1:-1] = (q - torch.roll(q, 1, dims=-1))[..., 1:]
+        p[..., -1] = 1 - q[..., -1]
```

Comment:
I'd try to replace these mutations with a purely functional form
```py
p = torch.cat([q[..., 0],
               (q - torch.roll(q, 1, dims=-1))[..., 1:],
               1 - q[..., -1]], dim=-1)
```
Actually I'm surprised the in-place edits don't break autograd. Are we sure gradient is tested?

Reply:
I think the gradient still works; there's like a `SelectBackward` (or `InsertBackward`? can't remember which) grad function for those types of in-place mutations. I double-checked that the gradient still flows through the log prob with the current implementation and it looks like it works:

```python
>>> from pyro.distributions import OrderedLogistic
>>> import torch
>>> pred = torch.randn(5, requires_grad=True)
>>> cp = torch.sort(torch.randn(3)).values
>>> cp.requires_grad = True
>>> data = torch.tensor([0,1,2,3,0]).float()
>>> dist = OrderedLogistic(pred, cp)
>>> dist.log_prob(data).sum().backward()
>>> pred.grad
tensor([-0.8494, -0.8638, -0.6831,  0.7186, -0.7939])
>>> cp.grad
tensor([-2.7713,  5.0584,  0.1845])
```
I was playing around with your suggested implementation and it seems that one problem is that if you only have a single cutpoint along the last dimension (i.e. two categories) you can get the following error:
```python
>>> q = torch.tensor([1.])
>>> torch.cat([q[...,0], (q-torch.roll(q, 1, dims=-1))[...,1:], 1-q[...,-1]], dim=-1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: zero-dimensional tensor (at position 0) cannot be concatenated
```
because the second element in the concat list is empty. But there's no problem assigning nothing to nothing
```python
>>> p = torch.tensor([0,0])
>>> p[...,1:-1] = (q-torch.roll(q, 1, dims=-1))[...,1:]
>>> p
tensor([0, 0])
```

Reply:
actually, that's probably a good idea to add to the tests!